File: dsa-problems/Queues/solutions/python/QUE-004-hallway-interleave.py
from typing import List
import sys

def interleave_queue(values: List[int]) -> List[int]:
    n = len(values)
    mid = n // 2
    result = []
    
    for i in range(mid):
        result.append(values[i])
        result.append(values[mid + i])
        
    return result

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    
    iterator = iter(input_data)
    try:
        n = int(next(iterator))
        values = [int(next(iterator)) for _ in range(n)]
        
        result = interleave_queue(values)
        print(" ".join(map(str, result)))
    except StopIteration:
        pass

if __name__ == "__main__":
    main()

File: dsa-problems/Queues/solutions/cpp/QUE-004-hallway-interleave.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> interleaveQueue(const vector<int>& values) {
        int n = values.size();
        int mid = n / 2;
        vector<int> result(n);
        
        for (int i = 0; i < mid; i++) {
            result[2 * i] = values[i];
            result[2 * i + 1] = values[mid + i];
        }
        
        return result;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    if (cin >> n) {
        vector<int> values(n);
        for (int i = 0; i < n; i++) {
            cin >> values[i];
        }
    
        Solution solution;
        vector<int> result = solution.interleaveQueue(values);
        for (int i = 0; i < (int)result.size(); i++) {
            if (i) cout << ' ';
            cout << result[i];
        }
        cout << "\n";
    }
    return 0;
}
